azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.4+ and Openssl 1.0+
+#
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.version import get_distro
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common import conf
+
+# Define support matrix for Regorus and policy engine feature.
+# Dict in the format: { distro:min_supported_version }
+POLICY_SUPPORT_MATRIX = {
+    'ubuntu': FlexibleVersion('16.04'),
+    'mariner': FlexibleVersion('1')
+}
+
+
+# Common logging function across PolicyEngineConfigurator and PolicyEngine classes,
+# so should be a module-level function.
+def log_policy(formatted_string, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+    """
+    Log information to console and telemetry.
+    """
+    if is_success:
+        logger.info("[Policy] " + formatted_string)
+    else:
+        logger.error("[Policy] " + formatted_string)
+    if send_event:
+        add_event(op=op, message=formatted_string, is_success=is_success)
+
+
+class PolicyEngineConfigurator:
+    """
+    This Singleton class is responsible for checking if policy is enabled for the agent.
+    Policy can only be enabled/disabled via configuration file, and can only be enabled on supported distros.
+    If an error occurs during Regorus import, policy will be globally disabled for the agent via the _policy_enabled flag.
+    Use: PolicyEngineConfigurator.get_instance()
+    """
+    _instance = None  # configurator is implemented as a singleton
+    _initialized = False  # set to true on first init, even if Regorus import fails, we don't want to retry.
+    _policy_enabled = False
+
+    def __init__(self):
+        # we only attempt to import Regorus if it is both enabled via conf and supported by the platform
+        try:
+            if PolicyEngineConfigurator._initialized:
+                return
+
+            if not conf.get_extension_policy_enabled():
+                log_policy("Policy enforcement is disabled via configuration file.")
+                return
+
+            if not self._is_policy_supported():
```

Comment:
In this case, we want to allow all extensions to be installed, but log that customer is trying to enable policy on unsupported distro.

Reply:
+1, the extension status message should be clear that extension processing is blocked because this is an unsupported distro (or whatever the reason is)

Reply:
Updated error handling to raise exceptions